pkg/geo/geomfn/subdivide.go

```diff
@@ -0,0 +1,364 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package geomfn
+
+import (
+	"math"
+
+	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/cockroachdb/errors"
+	"github.com/twpayne/go-geom"
+)
+
+// Subdivide returns a geometry divided into parts, where each part contains no more than the number of vertices provided.
+func Subdivide(g geo.Geometry, maxVertices int) (geo.Geometry, error) {
+	if g.Empty() {
+		return g, nil
+	}
+	const minMaxVertices = 5
+	if maxVertices < minMaxVertices {
+		return geo.Geometry{}, errors.Newf("max_vertices number cannot be less than %v", minMaxVertices)
+	}
+
+	gt, err := g.AsGeomT()
+	if err != nil {
+		return geo.Geometry{}, errors.Newf("could not transform input geometry into geom.T: %v", err)
+	}
+	col := geom.NewGeometryCollection()
+	dim, err := dimensionFromGeomT(gt)
+	if err != nil {
+		return geo.Geometry{}, errors.Newf("could not calculate geometry dimension: %v", err)
+	}
+	const startDepth = 0
+	err = subdivideRecursive(gt, maxVertices, startDepth, col, dim)
+	if err != nil {
+		return geo.Geometry{}, err
+	}
+	col.SetSRID(gt.SRID())
+	output, err := geo.MakeGeometryFromGeomT(col)
+	if err != nil {
+		return geo.Geometry{}, errors.Newf("could not transform output geom.T into geometry: %v", err)
+	}
+	return output, nil
+}
+
+func subdivideRecursive(
+	gt geom.T, maxVertices int, depth int, col *geom.GeometryCollection, dim int,
+) error {
+	const geometryCollectionPushError = "could not add geometry to results collection: %v"
+	// maxDepth 50 => 2^50 ~= 10^15 subdivisions.
+	const maxDepth = 50
+	clip := geo.BoundingBoxFromGeomTGeometryType(gt)
+	width := clip.HiX - clip.LoX
+	height := clip.HiY - clip.LoY
+	if width == 0 && height == 0 {
+		if gt, ok := gt.(*geom.Point); ok && dim == 0 {
+			err := col.Push(gt)
+			if err != nil {
+				return errors.Newf(geometryCollectionPushError, err)
+			}
+		}
+		return nil
+	}
+	// Boundaries needs to be widen slightly for intersection to work properly.
+	tolerance := 1e-12
+	if width == 0 {
+		clip.HiX += tolerance
+		clip.LoX -= tolerance
+		width = 2 * tolerance
+	}
+	if height == 0 {
+		clip.HiY += tolerance
+		clip.LoY -= tolerance
+		height = 2 * tolerance
+	}
+	// Recurse into collections other than MultiPoint.
+	isCollection, err := isGeomTCollection(gt)
```

Comment:
i think you can use type assertions here, i.e.

```
if _, ok := gt.(*geom.GeometryCollection); ok {

}
```

remove the function defined

Reply:
Actually, it's not the same. It's a wrapper for `IsCollection` function that works for `Geometry` and returns `true` not only for `GeometryCollection` type, but also for `MultiPolygon`, `MultiPoint`, and `MultiLineString`. I understand this "collection" here can be confusing, but I could not come up with any better name.

Reply:
maybe `isMultiGeomTShape`